---
title: "Comparaison d'isolats bactériens par séquençage génomique complet"
subtitle: ""
author: "Equipe PHE3ID"
date: "`r format(Sys.time(), '%d %B %Y')`"
params:
  cluster_id: ""
  species: ""
  replicon: "Chromosome"
  env_samples: ""
  path_to_reports: "2024-01-25_clone_reports.Rdata"
  path_to_scan:  "2024-01-25_annotation_scan.Rdata"
output:
  pdf_document:
      toc: true
      toc_depth: 2
      number_sections: true
fontsize: 12pt
#classoption: landscape
header-includes:
- \usepackage{graphicx}
- \usepackage{fancyhdr}
- \usepackage{float}
- \usepackage{titling}
- \usepackage{svg}
- \usepackage{booktabs}
- \usepackage[table]{xcolor}
- \pagestyle{fancy}
- \fancyhead[L]{Equipe PHE3ID}
- \fancyhead[C]{Cluster `r params$cluster_id` - `r params$replicon` - `r params$species`}
- \fancyhead[R]{\includegraphics[width=30pt]{pictures/logo_hcl.png}}
- \setlength{\headheight}{35pt}
- \addtolength{\topmargin}{-19pt}
- \pretitle{\begin{center}
  \hfill
  \includegraphics[width=1in,height=1in]{pictures/logo_hcl.png}\LARGE\\}
- \posttitle{\end{center}}
---

\newpage

Ce rapport d'études de cluster étudie les isolats de l'espèce ___ XXX ___ du service YYYY.

```{r setup, include=FALSE}
# Some cleaning [no: if cleaning, then we cannot use params data]
# rm(list = ls())

# Setting R chunks options
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.align = 'center')

# Loading libraries
usePackage <- function(p) {
  if (!is.element(p, installed.packages()[,1]))
    install.packages(p, dep = TRUE)
  require(p, character.only = TRUE)
}
libraries <- c("lubridate", "data.table", "stringr", "knitr", "ggplot2", "dplyr", "scales", "tidyr", "kableExtra", "ggpubr", "RColorBrewer", "tidyverse")
for (lib in libraries){usePackage(lib)}


# Loading Data
load(params$path_to_reports)
if(params$cluster_id==2){
  reports <- final_data[cluster_id %in% c(2, 161)]
} else if(params$cluster_id==348){
  reports <- final_data[cluster_id %in% c(348, 359)]
} else if(params$cluster_id==316){
  reports <- final_data[cluster_id %in% c(316, 322)]
} else {
  reports <- final_data[cluster_id == params$cluster_id]
}

# Studied species
if(params$species=="chromosome"){
  species <- "-"
} else {
  species <- params$species
}
# In case of ecloacae, include cronobacter !
if(params$species=="ecloacae"){
  # Also include cronobacter
  reports <- reports[mlst_species == species | mlst_species == "cronobacter",]
} else if(params$species=="ecoli"){
  # Also include ecoli_achtman_4
  reports <- reports[mlst_species == species | mlst_species == "ecoli_achtman_4",]
} else if(params$species=="kpneumoniae"){
  # Also include klebsiella
  reports <- reports[mlst_species == species | mlst_species == "klebsiella",]
} else {
  reports <- reports[mlst_species == species,]
}
rm(final_data)
reports <- reports %>%
  filter(genome!="Epi-205" | genome != "Epi-615" | genome != "Epi-2348")
# Global ggplot theme settings
global_theme <- theme(plot.title=element_text(size=16, face="plain", hjust = 0.5, vjust=2),
            axis.text.x=element_text(size=9),
            axis.text.y=element_text(size=10),
            axis.title.x=element_text(size=12, vjust = -1),
            axis.title.y=element_text(size=12, angle = 90, vjust = 2),
            legend.title=element_text(size=12),
            legend.text = element_text(size=9))
theme_set(global_theme)

rm(lib, libraries, usePackage)
```

# Description des prélèvements

```{r global_infos}
# Calcul de la date de prélèvement la + précoce
earliest <- format(min(reports$sampling_date), "%d/%m/%Y")
# Calcul de la date de prélèvement la + tardive
latest <- format(max(reports$sampling_date), "%d/%m/%Y")
# Nombre d'échantillons dans le cluster
nb_samples <- length(reports$sampling_date)
# Nombre de patients concernés
nb_patients <- length(unique(reports$ipp))
# Nombre d'échantillons environnement s'il y en a
nb_env_samples <- 0
if(!(params$env_samples=="")){
  nb_env_samples <- length(reports[reports$ipp == params$env_samples]$genome)
  nb_patients <- nb_patients - 1
}

```

```{r general_desc, results='asis'}
if(nb_patients==1){
  nb_patients <- paste(nb_patients, " patient")
} else {
  nb_patients <- paste(nb_patients, "patients")
}

if(nb_env_samples != 0){
  if(earliest==latest){
    cat("Groupe de ",nb_samples," isolats suspects chez ",nb_patients," et ",nb_env_samples," prélèvements d'environnement, prélevés le ", earliest, ".")
  } else {
    cat("Groupe de ",nb_samples," isolats suspects chez ",nb_patients," et ",nb_env_samples," prélèvements d'environnement, prélevés du ", earliest," au ",latest,".")
  }
} else {
  if(earliest==latest){
      cat("Groupe de ",nb_samples," isolats suspects chez ",nb_patients,", prélevé(s) le ", earliest, ".")
  } else {
      cat("Groupe de ",nb_samples," isolats suspects chez ",nb_patients,", prélevé(s) du ", earliest," au ",latest,".")
  }


}
rm(nb_env_samples, nb_samples, nb_patients, earliest, latest)
```
```{r dates, out.height="500pt"}
# Computing limits of histogram (earliest and latest months)
first_month <- floor_date(min(reports$sampling_date), "month") %m-% months(1)
last_month <- ceiling_date(max(reports$sampling_date), "month")
months_period <- interval(ymd(first_month),ymd(last_month)) %/% months(1)
if(months_period < 6){
  first_month <- as.Date("2023-01-01")
  last_month <- as.Date("2024-01-01")
}

if(params$cluster_id == 3){
  species_in_title <- "S. marcescens"
} else if(params$cluster_id %in% c(96, 113, 121, 127)){
  species_in_title <- "S. capitis"
} else if(params$species == "cfreundii" & params$cluster_id == 39) {
  species_in_title <- "C. freundii/braakii/youngae"
} else if(params$species == "ecloacae" & params$cluster_id == 39) {
  species_in_title <- "E. cloacae/hormaechei"
} else if(params$cluster_id == 50){
  species_in_title <- "A. faecalis"
} else if(params$cluster_id == 48){
  species_in_title <- "B. fragilis"
} else if(params$cluster_id %in% c(60, 115)){
  species_in_title <- "A. ursingii"
} else if(params$cluster_id == 40){
  species_in_title <- "P. mirabilis"
} else if(params$cluster_id == 99){
  species_in_title <- "A. xylosidans"
} else if(params$cluster_id == 87){
  species_in_title <- "A. nosocomialis"
} else if(params$cluster_id == 46){
  species_in_title <- "A. baumannii"
} else if(params$cluster_id == 73){
  species_in_title <- "S. saprophyticus"
} else if(params$cluster_id == 75){
  species_in_title <- "S. sonnei"
} else if(params$cluster_id == 111){
  species_in_title <- "M. catharralis"
} else {
  species_in_title <- gsub("^([A-Z])", "\\1. ", str_to_title(params$species))
}
# If there are some environmental samples in the group
# Break data by month, e.g one bin one month
out <- reports %>%
  select(genome, sampling_date, ipp) %>%
  mutate(month = floor_date(sampling_date, "month")) %>%
  mutate(patients=ifelse(ipp==params$env_samples, "environnement", "patient")) %>%
  select(-ipp)
ggplot_title <- bquote(atop("Histogramme des dates de prélèvement", 
                            "de l'espèce"~italic(.(species_in_title))))
# Plotting as a barplot of month categories
g <- ggplot(data=out, aes(x=month, fill = patients)) +
   geom_bar(stat="count", aes(text = paste('Nombre: ', after_stat(count))), width = 20) +
   ggtitle(ggplot_title) +
   labs(x = "Date de prélèvement par mois", y = "Nombre d'échantillons", fill = "Type de prélèvement") +
  scale_x_date(date_breaks = "month", date_labels = "%m/%Y", date_minor_breaks = "month", 
               limits = c(first_month, last_month)) +
  scale_y_continuous(n.breaks = 5) +
  theme(axis.text.x=element_text(angle=45, vjust = 1, hjust = 1))

g

rm(first_month, last_month, out, g, species_in_title, months_period)
```
\newpage

# Etudes comparatives basées sur les SNPs

```{r which_dist}
# Working on second scale of SNPs[10 or 15 SNPs generally]
clust_dist <- sort(str_match(colnames(reports), "clust_dist_\\d+"))[2]
nb_snps <- str_extract(clust_dist, "\\d+")
```

Deux isolats sont clonaux s'ils sont distants de moins **de `r nb_snps` SNPs**. Les isolats clonaux sont regroupés en cluster.

## Bilan par clone identifié

```{r clone, out.height="500pt"}
out <- reports %>%
  mutate(ipp=ifelse(ipp==params$env_samples, "environnement", ipp)) %>%
  mutate(text_on_bar=ifelse(ipp=="environnement", "environnement", ""))

# Color editing
cbPalette <- colorRampPalette(brewer.pal(9, "Set1"))

# Barplot of number of samples per clone
g <- ggplot(data=out, aes(x=as.factor(eval(as.name(clust_dist))), fill = ipp)) +
      geom_bar(stat="count")+
      stat_count(geom = "text", colour = "white", size = 3.5, angle = 0,
                 aes(label = ipp),position=position_stack(vjust=0.5)) +
      ggtitle(str_replace(clust_dist, "clust_dist_", "Clustering Distance ")) +
      labs(x = "Clusters", y = "Nombre d'échantillons", fill = "Identifiant Patient (IPP)") +
      scale_y_continuous(n.breaks = 5)+
      scale_fill_manual(values=cbPalette(length(unique(reports$ipp))))+
      guides(fill=guide_legend(ncol=2))

g

rm(g, cbPalette)
```

```{r desc_clust, results='asis'}
clusters <- names(which(table(reports[[clust_dist]]) > 1))
samples_clustered <- out[which(out[[clust_dist]] %in% clusters)]
nb_clusters <- length(clusters)
if(nb_clusters==1){
  cat("Au seuil de ", nb_snps," SNPs, 1 cluster d'isolats est formé.\n\n")
} else if(nb_clusters==0){
  cat("Au seuil de ", nb_snps," SNPs, aucun cluster d'isolats n'est formé.\n\n")
} else {
  cat("Au seuil de ", nb_snps," SNPs, ", nb_clusters, " clusters d'isolats sont formés.\n\n")
}
for(cl in clusters){
  cl_samples <- out[which(out[[clust_dist]] == cl)]
  cat("* Le cluster ", cl, " : groupe de ", length(cl_samples$genome), " isolats prélevés ")
  if("environnement" %in% cl_samples$ipp){
    # If there are environmental samples
    nb_env <- length(cl_samples[cl_samples$ipp == "environnement"]$genome)
    cat("de l'environnement")
    if(length(grep("\\d+", cl_samples$ipp)) > 0) {cat(" et ")}
  }
  if(length(grep("\\d+", cl_samples$ipp)) > 0){
    # If there are samples from patients
    nb_patients <- length(unique(cl_samples[cl_samples$ipp != "environnement"]$ipp))
    if(nb_patients > 1){
      cat("de ", nb_patients, " patients")
    } else {
      cat("d'un patient")
    }
  }
  cat(".\n\n")
}

rm(out, samples_clustered, cl, clusters, nb_clusters)
```
\newpage

## Bilan par prélèvement

La description des isolats étudiés :

```{r final_table}
# Arranging data
tab <- reports %>%
  select(glims, ipp, sampling_date, type_prlvt, soc_species, sequence_type,
         as.name(clust_dist)) %>%
  arrange(eval(as.name(clust_dist))) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="SURV", "Surveillance de flore", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="ECBU", "Examen cytobactériologique des urines", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="BPO", "Broncho-pulmonaire", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="CATHE", "Cathéter", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="HEMOA", "Hémoculture Flacon A", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="HEMOC", "Hémoculture Flacon C", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="LIQBIO", "Liquide Biologique", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="MUCO", "Mucoviscidose", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="HEMOA", "Hémoculture Flacon A", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="ORLC", "Oto-rhino-laryngologie", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="OSTEOC", "Infection osteo-articulaire", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="SELLE", "Selle", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="SUPE", "Suppuration prélevée par écouvillon", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="SUPP", "Suppuration profonde", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(str_to_upper(type_prlvt)=="BMR-BHR", "Dépistage rectal", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="RECHSP", "Prélèvement environnemental", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(type_prlvt=="Surv selles néonat", "Surveillance selles néonatologie", type_prlvt)) %>%
  mutate(type_prlvt=ifelse(is.na(type_prlvt), "", type_prlvt))

if(!(params$env_samples=="")){
  tab <- tab %>%
    mutate(ipp=ifelse(ipp==params$env_samples, "Prélèvement d'environnement", ipp))
}
  

# Outputting table 
kable(tab, col.names = c("GLIMS", "Identifiant Patient (IPP)", "Date de prélèvement", 
                         "Type de prélèvement", "Espèce MALDI-ToF", "MLST Sequence Type",
                         "Cluster"),
      align = "llllccc", "latex", booktabs = TRUE) %>%
  kable_styling(latex_options = c("striped", "scale_down", "HOLD_position", "repeat_header")) %>%
  column_spec(4, width = "15em")

rm(tab)
```

# Conclusion du clustering

```{r conclusion, results="asis"}
cat(readLines('conclusion.txt'), sep = '\n')
```

\newpage

# Informations complémentaires

```{r only_2_samples_clone, include=TRUE, results='asis'}
# First thing to do, only keep clone with nb of samples > 2 
singletons <- names(which(table(reports[[ clust_dist ]]) == 1))
reports[[clust_dist]][which(reports[[clust_dist]] %in% singletons)] <- "singletons"

# Compute proportion of non-singleton/singletons
prop_singleton <- round(length(reports$genome[which(reports[[clust_dist]] == "singletons")]) * 100 / length(reports$genome), digits = 2) 
prop_non_singleton <- 100 - prop_singleton

# Compute number of SNPs between samples if it's two non-clonal samples
if((length(reports$genome) == 2) & (prop_non_singleton == 0)){
  nb_variants <- unique(as.numeric(reports$nb_snps))
  cat("Les deux souches analysées ne sont pas clonales, elles sont distantes de ", nb_variants, " SNPs.")
}

rm(singletons, nb_variants)
```

La proportion de singletons est de `r prop_singleton` %, contre `r prop_non_singleton` % d'isolats clonaux.

N.B: Un singleton correspond à un isolat qui n'appartient à aucun cluster, il n'est clonal avec aucun autre isolat.

## Temporalité des clones

```{r prlvt, out.height="500pt"}
# Computing limits of histogram (earliest and latest months)
first_month <- floor_date(min(reports$sampling_date), "month") %m-% months(1)
last_month <- ceiling_date(max(reports$sampling_date), "month")
months_period <- interval(ymd(first_month),ymd(last_month)) %/% months(1)
if(months_period < 6){
  first_month <- as.Date("2023-01-01")
  last_month <- as.Date("2024-01-01")
}

# Break data by month, e.g one bin one month 
out <- reports %>%
  select(genome, as.name(clust_dist), sampling_date) %>%
  mutate(month = floor_date(sampling_date, "month"))

#cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
#cbPalette <- c('#999999', colorRampPalette(c("lightblue", "red"))(3))
#cbPalette <- c("#D2C9CA", palette(rainbow(length(unique(reports[[ clust_dist ]]))-1, "RdYlBu")))
cbPalette <- c("#D2C9CA", brewer.pal(length(unique(reports[[ clust_dist ]]))-1, "RdYlBu"))
# Plotting as a barplot of month categories
if("singletons" %in% out[[clust_dist]]){
  g <- ggplot(data=out, aes(x=month, fill = relevel(as.factor(eval(as.name(clust_dist))), "singletons"))) +
  geom_bar(colour = "black", aes(text = paste0("Nombre: ", after_stat(count))), width = 20) +
  ggtitle("Histogramme de dates de prélèvement") +
  labs(x = "Date de prélèvement par mois", y = "Nombre d'échantillons", fill = "Cluster") +
  scale_x_date(date_breaks = "month", date_labels = "%m/%Y", date_minor_breaks = "month", 
               limits = c(first_month, last_month)) +
  scale_y_continuous(n.breaks = 5) +
  scale_fill_manual(values=cbPalette) +
  theme(axis.text.x=element_text(angle=45, vjust = 1, hjust = 1))
} else {
  g <- ggplot(data=out, aes(x=month, fill = as.factor(eval(as.name(clust_dist))))) +
  geom_bar(colour = "black", aes(text = paste0("Nombre: ", after_stat(count))), width = 20) +
  ggtitle("Histogramme de dates de prélèvement") +
  labs(x = "Date de prélèvement par mois", y = "Nombre d'échantillons", fill = "Cluster") +
  scale_x_date(date_breaks = "month", date_labels = "%m/%Y", date_minor_breaks = "month", 
               limits = c(first_month, last_month)) +
  scale_y_continuous(n.breaks = 5) +
  scale_fill_manual(values=cbPalette[-1]) +
  theme(axis.text.x=element_text(angle=45, vjust = 1, hjust = 1))
}

# if(length(out$genome) < 10){
#   g <- g +
#     scale_y_continuous(n.breaks = length(out$genome))
# } else {
#   g <- g +
#     scale_y_continuous(n.breaks = 10)
# }

g

rm(first_month, last_month, out, g, earliest, latest, nb_samples, prop_singleton, prop_non_singleton, cbPalette, months_period)
```
\newpage

## Classes d'AMR par cluster

Les gènes de résistance indiqués sont uniquement les gènes localisés sur le chromosome. Les gènes de résistance sur plasmide ne sont pas inclus dans ce rapport.

```{r amr_barplot_per_classes, include=FALSE}
# Creating a list of AMR class found per clone
amr_list <- reports %>%
  select(as.name(clust_dist), amr_classes) %>%
  mutate(amr_classes = tolower(amr_classes))
amr_list <- separate_longer_delim(amr_list, amr_classes, ",")

# Histogram of AMR_classes (aminoglycoside, phenicol, etc.)
g <- ggplot(data=amr_list, aes(x=amr_classes)) +
  geom_bar(stat="count", fill = "lightblue", aes(text = paste0("Count: ", after_stat(count)))) +
  ggtitle("AMR classes") +
  labs(x = "AMR class", y = "Count") +
  scale_y_continuous(breaks= pretty_breaks()) +
  theme(axis.text.x=element_text(angle=45, vjust = 1, hjust = 1))

g

rm(g, amr_list)
```

```{r new_amr}
# Loading Data with AMRFinder report [GLOBALLY]
load(params$path_to_scan)
rm(amrfinder_reports, contig_reports, mge_reports, mlst_reports, quast_reports, sourmash_reports)
```

```{r amr_class_clone, results='asis'}
clones <- sort(unique(reports[[clust_dist]]))
for(c in clones){
  if(c == "singletons"){
    cat("### Singletons \n\n")
    cat("Au sein des **singletons**, on retrouve ces gènes de résistances aux antibiotiques : \n\n")
  } else {
    cat("### Cluster ", c, " \n\n")
    cat("Au sein du **cluster ", c, ",** on retrouve ces gènes de résistances aux antibiotiques : \n\n")    
  }
  
  # Which genomes are in my clusters ?
  samples <- reports$genome[reports[[clust_dist]] == c]
  # Get AMR report of those genomes
  amr_report <- combined_amr_report[`Element type` == "AMR" & genome %in% samples]
  # Loading only replicon AMR genes
  list_arg <- amr_report[molecule_type == str_to_lower(params$replicon)]
  
  # Arranging AMR results
  out <- list_arg %>%
    select(genome, `Gene symbol`, `Sequence name`, Class, Subclass) %>%
    distinct() %>%
    group_by(`Gene symbol`, `Sequence name`, Class, Subclass) %>%
    mutate(Class = str_to_title(Class), Subclass = str_to_title(Subclass)) %>%
    summarise(Pourcentage=round(n()*100/length(samples), digits = 2)) %>%
    arrange(-Pourcentage) %>%
    mutate(Pourcentage=paste(Pourcentage, "%"))
  
  # Print table with AMR results
  d <- kable(out, col.names = c("Nom du gène", "Description", "Classe d'AMR", "Sous-classe d'AMR", "Pourcentage de présence dans le cluster"), 
             align = "llllr", "latex", booktabs = TRUE) %>%
    kable_styling(latex_options = c("striped", "scale_down", "HOLD_position")) %>%
    column_spec(5, width = "8em") %>%
    column_spec(2, width = "20em")
  print(d)
  
  cat("\\newpage \n\n")
}

rm(amr_report, combined_amr_report, list_arg, out, c, clones, samples, d)
```

```{r tree, echo=FALSE, results='asis'}
# out.height="300pt"
# samples_clustered <- reports[[clust_dist]][which(reports[[clust_dist]] != "singletons")]
# if(length(reports$genome) > 2){
#   cat("\\newpage \n\n")
#   cat("## Phylogénie \n\n")
#   cat("A partir de l'alignement des SNPs détectés dans les échantillons, un arbre phylogénétique du cluster est inféré.\n\n")
#   if(length(samples_clustered)>0){
#     cat("Les isolats clonaux sont coloriés d'une couleur spécifique.\n\n")
#   }
#   path_to_tree <- paste0("../phylogeny/cluster_",params$cluster_id,"/",params$species,
#                          "/phylogenetic_tree/clonalframeml/GTR_tree.jpeg")
#   i <- include_graphics(path_to_tree)
#   cat("\\begin{center}")
#   cat("\\includegraphics[width=5in,height=7.5in]{", path_to_tree, "}")
#   cat("\\end{center}")
#   rm(path_to_tree)
# }
# rm(out.height)
```
```{r cleaning}
rm(global_theme, params, reports, species, nb_snps, clust_dist, samples_clustered)
```






